sdk/src/main/java/com/akkaserverless/javasdk/eventsourcedentity/EventSourcedEntityEffect.java

```diff
@@ -0,0 +1,137 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.akkaserverless.javasdk.eventsourcedentity;
+
+import com.akkaserverless.javasdk.Effect;
+import com.akkaserverless.javasdk.Metadata;
+import com.akkaserverless.javasdk.Reply;
+import com.akkaserverless.javasdk.ServiceCall;
+import com.akkaserverless.javasdk.impl.reply.FailureReplyImpl;
+import com.akkaserverless.javasdk.impl.reply.ForwardReplyImpl;
+import com.akkaserverless.javasdk.impl.reply.MessageReplyImpl;
+import com.akkaserverless.javasdk.impl.reply.NoReply;
+import com.akkaserverless.javasdk.reply.FailureReply;
+import com.akkaserverless.javasdk.reply.ForwardReply;
+import com.akkaserverless.javasdk.reply.MessageReply;
+
+import java.util.function.Function;
+
+/**
+ * A return type to allow returning forwards or failures, and attaching effects to messages.
+ *
+ * @param <T> The type of the message that must be returned by this call.
+ */
+public abstract class EventSourcedEntityEffect<T> implements Effect<T> {
+
+  /**
+   * Construct the effect that is returned by the command handler. The effect describes next
+   * processing actions, such as emitting events and sending a reply.
+   *
+   * @param <T> The type of the message that must be returned by this call.
+   * @param <S> The type of the state for this entity.
+   */
+  public static class Builder<T, S> {
+    private Builder() {}
+
+    public Builder2<T, S> emitEvent(Object event) {
+      throw new UnsupportedOperationException("Not implemented yet"); // FIXME
+    }
+
+    /**
+     * Create a message reply.
+     *
+     * @param payload The payload of the reply.
+     * @return A message reply.
+     */
+    public MessageReply<T> message(T payload) { // FIXME rename to reply? Note thenReply
+      return message(payload, Metadata.EMPTY);
+    }
+
+    /**
+     * Create a message reply.
+     *
+     * @param payload The payload of the reply.
+     * @param metadata The metadata for the message.
+     * @return A message reply.
+     */
+    public MessageReply<T> message(T payload, Metadata metadata) {
+      return new MessageReplyImpl<>(payload, metadata);
+    }
+
+    /**
+     * Create a forward reply.
+     *
+     * @param serviceCall The service call representing the forward.
+     * @return A forward reply.
+     */
+    public ForwardReply<T> forward(ServiceCall serviceCall) {
+      return new ForwardReplyImpl<>(serviceCall);
+    }
+
+    /**
+     * Create a failure reply.
+     *
+     * @param description The description of the failure.
+     * @return A failure reply.
+     */
+    public FailureReply<T> failure(String description) {
```

Comment:
Shall we rename this to `error`? https://www.reactivemanifesto.org/glossary#Failure

```
public ErrorReply<T> error(String description)
```

Reply:
Makes sense

Reply:
error it is: b5cea76